src/track/beatgrid.cpp

```diff
@@ -245,18 +244,18 @@ bool BeatGrid::findPrevNextBeats(double dSamples,
 }
 
 
-BeatIterator* BeatGrid::findBeats(double startSample, double stopSample) const {
+std::unique_ptr<BeatIterator> BeatGrid::findBeats(double startSample, double stopSample) const {
     QMutexLocker locker(&m_mutex);
     if (!isValid() || startSample > stopSample) {
-        return NULL;
+        return std::unique_ptr<BeatIterator>();
     }
     // qDebug() << "BeatGrid::findBeats startSample" << startSample << "stopSample"
     //          << stopSample << "beatlength" << m_dBeatLength << "BPM" << bpm();
     double curBeat = findNextBeat(startSample);
     if (curBeat == -1.0) {
-        return NULL;
+        return std::unique_ptr<BeatIterator>();
     }
-    return new BeatGridIterator(m_dBeatLength, curBeat, stopSample);
+    return std::make_unique<BeatGridIterator>(m_dBeatLength, curBeat, stopSample);
```

Comment:
BeatIterator is an abstract class and could not returned by value. The 
interface is defined by the base class.

On 02/11/2016 09:06 AM, Daniel Schürmann wrote:
In src/track/beatgrid.cpp 
<https://github.com/mixxxdj/mixxx/pull/892#discussion_r52571750>:
> >      }
> > -    return new BeatGridIterator(m_dBeatLength, curBeat, stopSample);
> > +    return std::make_unique<BeatGridIterator>(m_dBeatLength, curBeat, stopSample);
>
> It feels weird to delete an Iterator, but I understand that how it 
> currently works.
>
> It is only used here
> https://github.com/mixxxdj/mixxx/blob/49aaf97de17612b57d2a7618fcae0e687dbea8e0/src/waveform/renderers/waveformrenderbeat.cpp#L52
> Cant we return the BeatGridIterator by value and rely on the Compilers 
> return value optimization that avoids a deep copy?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/892/files#r52571750>.
>